### ISSUE TITLE
Add CI job for the rpm-sequoia backend

### ISSUE THREAD

**Issue by nwalfield** [rpm-software-management/rpm#2065]
Now that the [Sequoia OpenPGP backend has been merged](https://github.com/rpm-software-management/rpm/pull/2043), it would be good to have a CI job to make sure the code doesn't bit rot.

Thanks to @decathorpe, there's already [some preliminary work on packaging rpm-sequoia] for Fedora.  (That work was based on v0.1.  He reported a few build issues and some things that needed to be changed to be in compliance with the Fedora.  Those issues should all be resolved now, and I release [0.2](https://crates.io/crates/rpm-sequoia).)

I'd propose using a separate job to increase the parallelism.

rpm uses Semaphore CI.  I've never used Semaphore.  I took a look and I couldn't figure out where CI jobs are described.  I'd appreciate any pointers.

**Comment by nwalfield** [comment]
It seems I need to modify ci/Dockerfile and it will magically be picked up.  I'll work on that.  Guidance is still appreciated.

**Comment by pmatilai** [comment]
Yeah most of the "logic" is in ci/Dockerfile. Testing non-default options is an issue on general level, not just this. I suppose we could add separate jobs from SemaphoreCI but we have no precedent for doing that.

Note that you can do 'make ci' to test-drive your changes locally, a bit less pain than pushing a PR and then waiting for semaphore queues and all.

**Comment by nwalfield** [comment]
Thanks for the tip!

**Comment by pmatilai** [comment]
semaphoreci job control is a web panel thing, I'm not aware them having anything else.

**Comment by pmatilai** [comment]
Well we'll need to decide how we want to handle this all. I don't quite have an opinion offhand.

**Comment by nwalfield** [comment]
Ok, I think I have something reasonable now.

The basic idea is when we build the docker image, we install the dependencies, fetch rpm-sequoia from git, build it, and then install it.  Then when we run the container the first argument is the crypto backend to use (either libgcrypt, sequoia, or openssl).  master only tests libgcrypt.  To also test Sequoia, we'd need a another job that executes the container a second time, and passes "sequoia".  These two jobs could run in parallel.  What do you think about this design?

**Comment by nwalfield** [comment]
This patch is now ready for review.

(`make distcheck` is rather tricky. It doesn't seem to pass through environment variables like `PKG_CONFIG_PATH` so it needs to be set as an argument to `make`, but `LD_LIBRARY_PATH` can only be set in the environment.)

Currently three tests are failing for the Sequoia backend:

```
271: rpmkeys --import, signed with an expired subkey FAILED (rpmsigdig.at:385)
272: rpmkeys --import, signed with a revoked subkey  FAILED (rpmsigdig.at:452)
273: rpmkeys --import invalid keys                   FAILED (rpmsigdig.at:524)
```

```
271. rpmsigdig.at:382: testing rpmkeys --import, signed with an expired subkey ...
../../tests/rpmsigdig.at:385:
...
--- /dev/null	2022-04-22 22:03:25.100325727 +0000
+++ .../rpm/b/tests/rpmtests.dir/at-groups/271/stderr	2022-05-13 09:55:15.677911158 +0000
@@ -0,0 +1,3 @@
+warning: Certificate B3A771BFEB04E625:
+  Subkey 1F71177215217EE0 is expired: The subkey is not live
+  Certificate does not have any usable signing keys
271. rpmsigdig.at:382: 271. rpmkeys --import, signed with an expired subkey (rpmsigdig.at:382): FAILED (rpmsigdig.at:385)
```

This is due to [the linting change](https://github.com/rpm-software-management/rpm/pull/2052).  That breakage clearly motivates the addition of this change :D.  I'll fix that.

**Comment by nwalfield** [comment]
Merging https://github.com/rpm-software-management/rpm/pull/2063 fixes the tests and the CI passes with the Sequoia backend.

When this PR is merged, please also create another CI job so that the rpm container is run once for the sequoia backend and once for the libgcrypt backend.  (The container only needs to be built once.)

**Comment by nwalfield** [comment]
@decathorpe has submitted packages for `rpm-sequoia` and a dependency that is not yet packaged for Fedora to [copr](https://copr.fedorainfracloud.org/coprs/decathorpe/sequoia-test-builds/packages/). It appears to [successfully build on all architectures](https://copr.fedorainfracloud.org/coprs/decathorpe/sequoia-test-builds/build/4417227).

This PR currently builds `rpm-sequoia` from source.  Using packages would take less time.  On the other hand, if we make changes to `rpm-sequoia`, then we won't be able to use those changes in `rpm` until we release a new version of `rpm-sequoia`, and it makes it into Fedora.  @pmatilai (and others): What do you think?

**Comment by pmatilai** [comment]
No reason we couldn't pull it from copr, I'd much rather have that than building from source.
A bigger showstopper is that we're currently stuck on Fedora 34 because of fakechroot problems in newer, with no easy way forward.

**Comment by nwalfield** [comment]
It seems that the latest version of [Sequoia is being shipped for Fedora 34](https://packages.fedoraproject.org/pkgs/rust-sequoia-openpgp/rust-sequoia-openpgp-devel/), so I guess that is / will be an option.  I don't know enough about copr: will it automatically try to build new packages for Fedora 34 or does it need to be triggered automatically?

**Comment by nwalfield** [comment]
I think it would be better to add a sequoia CI job sooner rather than later.  Once the Fedora package is available, we can switch.  What do you prefer, @pmatilai?

**Comment by decathorpe** [comment]
We have the latest sequoia-openpgp packages in all Fedora branches, including 34. If it helps you, I can enable the chroots for Fedora 34 in my COPR as well, until the packages are through review and imported to Fedora.

**Comment by nwalfield** [comment]
Thanks @decathorpe!

I've now updated the Dockerfile to use @decathorpe's build of `rpm-sequoia` on cop, so that the CI does not need to build `rpm-sequoia`.

**Comment by nwalfield** [comment]
@pmatilai : Is there anything that I should do to improve the proposed solution?  Can I help in any other way? Thanks.

**Comment by pmatilai** [comment]
Sorry about the silence, I've been on and off PTO for the last two weeks and haven't managed that much serious thought in the middle :sweat_smile: 

I don't see anything particularly wrong with the proposed solution as such. However to me, this is just a part of a bigger, more generic question: what can we afford to test in the CI, and various practical questions how to go about it. I don't want to send you on a fool's errand one way or the other until I have some answers. So to be clear, this is blocking on me, not your work.

**Comment by nwalfield** [comment]
Thanks for the clarification.  I'm happy to let you do the thinking or to collaborate with you on that, perhaps using a high-bandwidth medium (audio / video chat) if you like.

**Comment by nwalfield** [comment]
Hi, [over here](https://github.com/rpm-software-management/rpm/issues/2141#issuecomment-1235120698), @pmatilai wrote:

> The ci part will have to wait for another release though.

I think it would be good to figure this out sooner rather than later: it would be a shame if a commit broke the Sequoia backend, just because we don't have a CI job.  @pmatilai, could you share your thinking about CI.  Do your reservations have to do with a financial cost?

**Comment by pmatilai** [comment]
It's cost, but in a wider meaning than just dollars and euros. We're currently only testing the default setup, and going beyond that opens an unpleasant box of permutations that we'd need to test just as desperately, including but not limited to the different rpmdb backends. The more material cost is SemaphoreCI quota, we've been considering moving to GH Actions instead to have it all here (resistance seems futile, if you catch my drift :sweat_smile: )

So what I'm currently thinking is that once the dust settles a bit, we'll just switch the Sequoia to be the default on CI and punt the permutations issue again.

**Comment by nwalfield** [comment]
I understand, thanks for the explanation.  Note: if we switch to GitHub actions, there is still the 2k minutes per month quota.

Btw, I [changed rpm-sequoia's CI from gitlab's to GitHub actions and it was pretty straightforward](https://github.com/rpm-software-management/rpm-sequoia/blob/main/.github/workflows/ci.yml).  Even the caching is easy to use.  One problem for rpm is that [GitHub doesn't have Fedora images](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources) out of the box.

**Comment by pmatilai** [comment]
Answering the question in #2272 here to keep things in context:

> Has the dust settled? As the sequoia backend will be the default in Fedora 38, is it time to use it in CI?

Getting warmer, but not there yet. There's just too many moving parts on too many different fronts right now. Getting it actually enabled and running for a few weeks without major issues in rawhide is a milestone after we can ask to build rpm-sequoia for the by then released F37 too, and then we can look at updating CI to use that and then swap to Sequoia for CI. (and getting this puzzle out of my head into writing can't hurt)

**Comment by pmatilai** [comment]
Seems decathorpe made rpm-sequoia available even on F36, but as F37 was recently released I just updated our CI to that in #2295. rpm-sequoia 1.2.0 was submitted to F37 stable updates a moment ago and I've verified that the test-suite passes with that, so we should be able to swap to this next week. Getting there!

**Comment by decathorpe** [comment]
Yeah, I submitted builds for rpm-sequoia to stable branches as well to make testing / CI easier. :) Hope it helps.
